File: lib/rgfa/field/byte_array.rb
module RGFA::Field::ByteArray

  def unsafe_decode(string)
    string.to_byte_array(valid: true)
  end

  def decode(string)
    string.to_byte_array
  end

  def validate_encoded(string)
    if string !~ /^[0-9A-F]+$/
      raise RGFA::FormatError,
        "#{string.inspect} is not a valid hex string\n"+
        "(it does not match the regular expression [0-9A-F]+)"
    end
  end

  def validate_decoded(byte_array)
    byte_array.validate
  end

  def unsafe_encode(object)
    case object
    when RGFA::ByteArray
      object.to_s(valid: true)
    when Array
      object.to_byte_array.to_s(valid: true)
    when String
      object
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, RGFA::ByteArray, Array)"
    end
  end

  def encode(object)
    case object
    when RGFA::ByteArray
      object.to_s
    when Array
      object.to_byte_array.to_s
    when String
      validate_encoded(object)
      object
    else
      raise RGFA::TypeError,
        "the class #{object.class} is incompatible with the datatype\n"+
        "(accepted classes: String, RGFA::ByteArray, Array)"
    end
  end

  module_function :decode
  module_function :unsafe_decode
  module_function :validate_encoded
  module_function :validate_decoded
  module_function :unsafe_encode
  module_function :encode

end

File: gfapy/field/byte_array.py
import gfapy
import re

def unsafe_decode(string):
  return gfapy.ByteArray(string)

def decode(string):
  return gfapy.ByteArray(string)

def validate_encoded(string):
  if not re.match(r"^[0-9A-F]+$", string):
    raise gfapy.FormatError(
      "{} is not a valid hex string\n".format(repr(string))+
      "(it does not match the regular expression [0-9A-F]+)")

def validate_decoded(byte_array):
  return byte_array.validate()

def unsafe_encode(obj):
  if isinstance(obj, gfapy.ByteArray):
    return str(obj)
  if isinstance(obj, list):
    return str(gfapy.ByteArray(obj))
  elif isinstance(obj, str):
    return obj
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, list, gfapy.ByteArray)")

def encode(obj):
  if isinstance(obj, gfapy.ByteArray):
    return str(obj)
  elif isinstance(obj, list):
    return str(gfapy.ByteArray(obj))
  elif isinstance(obj, str):
    validate_encoded(obj)
    return obj
  else:
    raise gfapy.TypeError(
      "the class {} is incompatible with the datatype\n"
      .format(obj.__class__.__name__)+
      "(accepted classes: str, list, gfapy.ByteArray)")
